src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,350 @@
+
+(defn- unify-vars
+  "Replace all template vars in `form` with references to corresponding models
+   in `context` as returned by `bind-models`.
+   `form` can be either a map (MBQL) or string (native SQL).
+   Aborts if any of the template vars cannot be matched."
+  [context form]
+  (try
+    (if (map? form)
+      (postwalk
+       (fn [subform]
+         (if (template-var? subform)
+           (or (some->> subform
+                        (unify-var context)
+                        ->reference)
+               (throw (Throwable.)))
+           subform))
+       form)
+      (s/replace form #"\?\w+" (fn [token]
+                                 (or (some->> token (unify-var context) :name)
+                                     (throw (Throwable.))))))
+    (catch Throwable _ nil)))
+
+(def ^:private ^Integer grid-width 18)
+(def ^:private ^Integer card-width 6)
+(def ^:private ^Integer card-height 4)
+
+(defn- next-card-position
+  "Return `:row` x `:col` coordinates for the next card to be placed on
+   dashboard `dashboard`.
+   Assumes a grid `grid-width` cells wide with cards sized
+   `card-width` x `card-height`."
+  [dashboard]
+  (let [num-cards (db/count 'DashboardCard :dashboard_id (:id dashboard))]
+    {:row (int (* (Math/floor (/ (* card-width num-cards)
+                                 grid-width))
+                  card-height))
+     :col (int (* (Math/floor (/ (mod (* card-width num-cards) grid-width)
+                                 card-width))
+                  card-width))}))
+
+(defn- create-dashboards!
+  [dashboards]
+  (into {}
+    (for [{:keys [title as descriptioin]} dashboards]
+      [as (delay (let [dashboard (db/insert! Dashboard
+                                   :name        title
+                                   :description descriptioin
+                                   :creator_id  api/*current-user-id*
+                                   :parameters  [])]
+                   (events/publish-event! :dashboard-create dashboard)
+                   dashboard))])))
+
+(defn- add-to-dashboard!
+  [dashboard card]
+  (dashboard/add-dashcard! dashboard card (merge (next-card-position dashboard)
+                                                 {:sizeX card-width
+                                                  :sizeY card-height}))
+  dashboard)
+
+(defn- build-query
+  [database context query]
+  (when-let [query (unify-vars context query)]
+    (let [database      (if (and (map? query)
+                                 (string? (:source_table query))
+                                 (s/starts-with? (:source_table query) "card__"))
+                          virtual-id
+                          database)
+          dataset_query (if (map? query)
+                          {:type     :query
+                           :query    query
+                           :database database}
+                          {:type     :native
+                           :native   {:query query}
```

Comment:
this reminds me how much I _hate_ having `:native` as the map key instead of just having a generic `:query` map key for all dataset query maps.

Reply:
@salsakran it's been like that since before I joined. What was the reasoning behind doing things that way? I never thought it made a ton of sense either TBH